neps/nep-0393.md

```diff
@@ -0,0 +1,516 @@
+
+```rust
+trait SBTNFT {
+    fn nft_total_supply(&self) -> U64;
+    // here we index by token id instead of by class id (as done in `sbt_tokens_by_owner`)
+    fn nft_tokens_for_owner(&self, account_id: AccountId, from_index: Option<U64>, limit: Option<u64>) -> Vec<Token>;
+    fn nft_supply_for_owner(&self, account_id: AccountId) -> U64;
+}
+```
+
+### Events
+
+Event design principles:
+
+- Events don't need to repeat all function arguments - these are easy to retrieve by indexer (events are consumed by indexers anyway).
+- Events must include fields necessary to identify subject matters related to use case.
+- When possible, events should contain aggregated data, with respect to the standard function related to the event.
+
+```typescript
+// only valid integer numbers (without rounding errors).
+type u64 = number;
+
+type Nep393Event {
+  standard: "nep393";
+  version: "1.0.0";
+  event: "mint" | "recover" | "renew" | "revoke" | "burn" | "ban" | "soul_transfer" ;
+  data: Mint | Recover | Renew | Revoke | Burn | Ban[] | SoulTransfer;
+}
+
+/// An event minted by the Registry when new SBT is created.
+type Mint {
+  ctr: AccountId;    // SBT Contract minting the tokens
+  tokens: [[AccountId, u64[]]];  // list of pairs (token owner, TokenId[])
+}
+
+/// An event emitted when a recovery process succeeded to reassign SBTs, usually due to account
+/// access loss. This action is usually requested by the owner, but executed by an issuer,
+/// and doesn't trigger Soul Transfer. Registry reassigns all tokens assigned to `old_owner`
+/// that were ONLY issued by the `ctr` SBT Contract (hence we don't need to enumerate the
+/// token IDs).
+/// Must be emitted by an SBT registry.
+type Recover {
+  ctr: AccountId         // SBT Contract recovering the tokens
+  old_owner: AccountId;  // current holder of the SBT
+  new_owner: AccountId;  // destination account.
+}
+
+/// An event emitted when a existing tokens are renewed.
+/// Must be emitted by an SBT registry.
+type Renew {
+  ctr: AccountId;  // SBT Contract renewing the tokens
+  tokens: u64[];  // list of token ids.
+}
+
+/// An event emitted when an existing tokens are revoked.
+/// Revoked tokens will continue to be listed by the registry but they should not be listed in
+/// a wallet. See also `Burn` event.
+/// Must be emitted by an SBT registry.
+type Revoke {
+  ctr: AccountId;  // SBT Contract revoking the tokens
+  tokens: u64[];  // list of token ids.
+}
+
+/// An event emitted when an existing tokens are burned and removed from the registry.
+/// Must be emitted by an SBT registry.
+type Burn {
+  ctr: AccountId;  // SBT Contract burning the tokens
+  tokens: u64[];  // list of token ids.
+}
+
+/// An event emitted when the `account` is banned within the emitting registry.
+/// Registry must add the `account` to a list of accounts that are not allowed to get any SBT
+/// in the future.
+/// Must be emitted by an SBT registry.
+type Ban {
+  account: AccountId;
+}
```

Comment:
Would this ever be emitted without a `SoulTransfer`? If not, does having a separate event make sense?

Reply:
> Would a soul transfer issue two events, Ban and SoulTransfer

yes

> should a SoulTransfer also be treated as a Ban?

No. These events are not the same. Ban can be issued in other occasions, mentioned in the NEP.

Reply:
But would a `SoulTransfer` ever not include a `Ban`? If so, why not enforce it by the protocol

Reply:
SoulTransfer emits `Ban` event:
> Whenever a soul transfer is triggered then the SBT registry emits `SoulTransfer` and `Ban` events